lib/core.ml

```diff
@@ -279,25 +279,50 @@ let match_str ~groups ~partial re s ~pos ~len =
   in
   let initial_state = find_initial_state re initial_cat in
   let st = scan_str info s initial_state ~groups in
+  (* This function adds a final boundary check on the input.
+     This is useful to indicate that the output failed because
+     of insufficient input, or to verify that the output actually
+     matches for regex that have boundary conditions with respect
+     to the input string.
+   *)
+  let finalize () =
```

Comment:
We don't call this function in every branch, but we allocate the closure eagerly. Can we hoist it up outside of this function?

Reply:
Sorry for the delay, somehow I missed the notification for this, and I went back to check on it, and found that there was a response. I pushed a new version that hoists this function.